bulkwebhook/bulk_webhook/doctype/kafka_hook/kafka_hook.py

```diff
@@ -98,62 +140,31 @@ def run_webhooks(doc, method):
         frappe.flags.kafkahook_executed = {}
 
     if frappe.flags.kafkahook is None:
-        # load webhooks from cache
-        webhooks = frappe.cache().get_value("kafkahook")
-        if webhooks is None:
-            # query webhooks
-            webhooks_list = frappe.get_all(
-                "Kafka Hook",
-                fields=["name", "`condition`", "webhook_docevent", "webhook_doctype"],
-                filters={"enabled": True},
-            )
-            # make webhooks map for cache
-            webhooks = {}
-            for w in webhooks_list:
-                webhooks.setdefault(w.webhook_doctype, []).append(w)
-            frappe.cache().set_value("kafkahook", webhooks)
-
-        frappe.flags.kafkahook = webhooks
+        frappe.flags.kafkahook = fetch_webhooks_from_redis()
 
-    # get webhooks for this doctype
-    webhooks_for_doc = frappe.flags.kafkahook.get(doc.doctype, None)
+    webhooks_for_doc = frappe.flags.kafkahook.get(doc.doctype)
 
     if not webhooks_for_doc:
-        # no webhooks, quit
         return
 
-    def _webhook_request(webhook):
-        if webhook.name not in frappe.flags.kafkahook_executed.get(doc.name, []):
-            frappe.enqueue(
-                "bulkwebhook.bulk_webhook.doctype.kafka_hook.kafka_hook.enqueue_webhook",
-                enqueue_after_commit=True,
-                doc=doc,
-                kafka_hook=webhook,
-            )
-
-            # keep list of webhooks executed for this doc in this request
-            # so that we don't run the same webhook for the same document multiple times
-            # in one request
-            frappe.flags.kafkahook_executed.setdefault(doc.name, []).append(
-                webhook.name
-            )
-
     event_list = ["on_update", "after_insert", "on_submit", "on_cancel", "on_trash"]
 
+    # value change is not applicable in insert
     if not doc.flags.in_insert:
-        # value change is not applicable in insert
         event_list.append("on_change")
         event_list.append("before_update_after_submit")
 
-    from frappe.integrations.doctype.webhook.webhook import get_context
-
     for webhook in webhooks_for_doc:
         trigger_webhook = False
         event = method if method in event_list else None
+
         if not webhook.condition:
             trigger_webhook = True
-        elif frappe.safe_eval(webhook.condition, eval_locals=get_context(doc)):
+        elif frappe.safe_eval(
+            webhook.condition, eval_locals={**WEBHOOK_CONTEXT, "doc": doc}
+        ):
             trigger_webhook = True
 
         if trigger_webhook and event and webhook.webhook_docevent == event:
-            _webhook_request(webhook)
+            if webhook.name not in frappe.flags.kafkahook_executed.get(doc.name, []):
```

Comment:
- Can we put this first? Feels like we should run eval later.
- Flat > Nested.

Reply:
Agreed. Had overlooked this bit because it didn't seem like a "big enough" problem in comparison. Updated via https://github.com/aakvatech/Bulk-Webhook/pull/1/commits/3068e8d335ecdfc6543a2aa7dc0d18ef04c0160a